### ISSUE TITLE
One suggestion may be useful for learning

### ISSUE THREAD

**Issue by ZY-WANG-0304** [carpedm20/BEGAN-pytorch#7]
Hi! Refering your code, I make below changes in my code.

![1](https://cloud.githubusercontent.com/assets/8197085/25927965/23eecfaa-362c-11e7-896c-52df9daec6c4.png)
![2](https://cloud.githubusercontent.com/assets/8197085/25927982/33134b14-362c-11e7-920b-b2b55808e9d7.png)
![3](https://cloud.githubusercontent.com/assets/8197085/25927995/3ee6527e-362c-11e7-849b-df5350dc3f92.png)

I change the way of the initializing at first.
Then, I change the way to get the loss. In my code, the both part of the g_loss would be used for learning G's parameters, but in your code it maybe only the AE_G be used for learning G's parameters.
I'm not sure about this, but maybe you can try it. And it's not big changes.
I'm sorry for my pool English~

**Comment by carpedm20** [comment]
Thanks for sharing your code. Did you tried to train the model with your code? If so, how does the samples look like?

**Comment by nian-liu** [comment]
@carpedm20 Actually, here is another PyTorch implementation for BEGAN: https://github.com/sunshineatnoon/Paper-Implementations/tree/master/BEGAN. Although that code lacks two tanh layer at the ends of G and D networks, I found it works well after adding two tanh. After coarsely comparing that code with this one, I found that code strictly obeys the basic training scheme of GANs while this code does not. Maybe some senior tricks are used in this code that I can't undetstand, but you can try this:
1.   Use d_loss and g_loss to separately and successively update D and G instead of adding them up and updating the whole network once. (line 166 and 167 in trainer.py)
2. Use g_loss = l1(AE_G_g, sample_z_G.detach()) on line 164

**Comment by ZY-WANG-0304** [comment]
![500_g](https://cloud.githubusercontent.com/assets/8197085/25930324/37e22604-3638-11e7-8822-11148aa203dd.png)
![5000_g](https://cloud.githubusercontent.com/assets/8197085/25930329/3e0ab488-3638-11e7-97ef-424c8e99e089.png)
![8000_g](https://cloud.githubusercontent.com/assets/8197085/25930335/428b98c4-3638-11e7-85d0-c7e08d6ba240.png)
![17500_g](https://cloud.githubusercontent.com/assets/8197085/25930338/4a40d2e6-3638-11e7-8b8b-bb6a4ff3c822.png)
Here is some samples I generated. 500_G,5000_G,8000_G & 17500_G. I change the lr from 0.00008 to 0.00004 at 5000 iters and to 0.00002 at 10000iters.
I'm sure that I'm using non-optimal parameters and super-parameters. And I don't consider the measure when I update the lr.

**Comment by ZY-WANG-0304** [comment]
@nian-liu I've seen the code you refered. Maybe he has changed his code,  the code to compute loss is similiar to mine.
![image](https://cloud.githubusercontent.com/assets/8197085/25930937/8fc0ad2a-363b-11e7-8a0c-99c03b137e8d.png)
And I guess the tanh() is helpful for solving pytorch's bad initialization. I'm not sure, just for communication.
